BuildAndTest.proj

```diff
@@ -94,9 +94,74 @@
+      <ProcDumpPath Condition="'$(ProcDumpPath)' == ''">C:\Sysinternals\Procdump.exe</ProcDumpPath>
+      <ProcessWatchdogArgs>--output-folder $(ProcessWatchdogOutputDirectory) --screenshot --procdump-path "$(ProcDumpPath)"</ProcessWatchdogArgs>
+      <BufferSeconds>300</BufferSeconds>
+      <!--
+      We'd have liked to extract these constants from DateTime, for example,
+          $([System.DateTime]::TicksPerSecond)
+      but MSBuild won't let you do that, because MSBuild lets you invoke methods and
+      properties, but TicksPerSecond is not a property, it's a constant.
+      -->
+      <TicksPerSecond>10000000</TicksPerSecond>
+      <TicksPerMinute>$([MSBuild]::Multiply($(TicksPerSecond), 60))</TicksPerMinute>
+
+      <BufferTicks>$([MSBuild]::Multiply($(BufferSeconds), $(TicksPerSecond)))</BufferTicks>
+
+      <StartTicks>$([System.DateTime]::Parse($(BuildStartTime)).Ticks)</StartTicks>
+      <NowTicks>$([System.DateTime]::Now.Ticks)</NowTicks>
+      <TimeSinceStartTicks>$([MSBuild]::Subtract($(NowTicks), $(StartTicks)))</TimeSinceStartTicks>
+      <BuildTimeLimitTicks>$([MSBuild]::Multiply($(BuildTimeLimit), $(TicksPerMinute)))</BuildTimeLimitTicks>
+      <RemainingTimeTicks>$([MSBuild]::Subtract($([MSBuild]::Subtract($(BuildTimeLimitTicks), $(TimeSinceStartTicks))), $(BufferTicks)))</RemainingTimeTicks>
+      <RemainingTimeSeconds>$([MSBuild]::Divide($(RemainingTimeTicks), $(TicksPerSecond)))</RemainingTimeSeconds>
+    </PropertyGroup>
+
+    <Error Condition="'$(RunProcessWatchdog)' == 'true' And $(RemainingTimeSeconds) &lt;= 0" Text="There is no time remaining to run the tests." />
+    <Message Condition="'$(RunProcessWatchdog)' == 'true'" Importance="high" Text="Time remaining to run tests = $(RemainingTimeSeconds) seconds." />
+
+    <Exec Condition="'$(RunProcessWatchdog)' == 'true'" Command='$(ProcessWatchdogExe) --executable $(CoreRunExe) --arguments "$(CoreRunArgs)" --time-limit $(RemainingTimeSeconds) $(ProcessWatchdogArgs)' />
+
+    <!-- Recalculate the remaining time after the first set of tests. -->
+    <PropertyGroup Condition="'$(RunProcessWatchdog)' == 'true'">
+      <NowTicks>$([System.DateTime]::Now.Ticks)</NowTicks>
+      <TimeSinceStartTicks>$([MSBuild]::Subtract($(NowTicks), $(StartTicks)))</TimeSinceStartTicks>
+      <BuildTimeLimitTicks>$([MSBuild]::Multiply($(BuildTimeLimit), $(TicksPerMinute)))</BuildTimeLimitTicks>
+      <RemainingTimeTicks>$([MSBuild]::Subtract($([MSBuild]::Subtract($(BuildTimeLimitTicks), $(TimeSinceStartTicks))), $(BufferTicks)))</RemainingTimeTicks>
+      <RemainingTimeSeconds>$([MSBuild]::Divide($(RemainingTimeTicks), $(TicksPerSecond)))</RemainingTimeSeconds>
+    </PropertyGroup>
+    
+    <Error Condition="'$(RunProcessWatchdog)' == 'true' And $(RemainingTimeSeconds) &lt;= 0" Text="There is no time remaining to run the tests." />
```

Comment:
We should move the logic here into a powershell script.  I'd prefer MSBuild scripts to be simple as possible as they're difficul/t to debug.